Common/BinaryComparisonExtensions.cs

```diff
@@ -0,0 +1,188 @@
+            TKey reference
+            )
+        {
+            var result = new SortedDictionary<TKey, TValue>();
+            if (comparison.Type == ExpressionType.Equal)
+            {
+                TValue value;
+                if (values.TryGetValue(reference, out value))
+                {
+                    result.Add(reference, value);
+                }
+
+                return result;
+            }
+
+            // since we're enumerating a sorted collection, once we receive
+            // a mismatch it means we'll never again receive a match
+            var breakAfterFailure =
+                comparison == BinaryComparison.LessThanOrEqual ||
+                comparison == BinaryComparison.LessThanOrEqual;
+
+            var evaluator = comparison.GetEvaluator<TKey>();
+            foreach (var kvp in values)
+            {
+                if (evaluator(kvp.Key, reference))
+                {
+                    result.Add(kvp.Key, kvp.Value);
+                }
+                else if (breakAfterFailure)
+                {
+                    break;
+                }
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Filters the provided <paramref name="values"/> according to this <see cref="BinaryComparison"/>
+        /// and the specified <paramref name="reference"/> value. The <paramref name="reference"/> value is
+        /// used as the RIGHT side of the binary comparison. Consider the binary comparison is LessThan and
+        /// we call Filter(values, 42). We're looking for keys that are less than 42.
+        /// </summary>
+        public static ImmutableSortedDictionary<TKey, TValue> Filter<TKey, TValue>(
+            this BinaryComparison comparison,
+            ImmutableSortedDictionary<TKey, TValue> values,
+            TKey reference
+            )
+        {
+            if (comparison.Type == ExpressionType.NotEqual)
+            {
+                return values.Remove(reference);
+            }
+
+            var result = ImmutableSortedDictionary<TKey, TValue>.Empty;
+            if (comparison.Type == ExpressionType.Equal)
+            {
+                TValue value;
+                if (values.TryGetValue(reference, out value))
+                {
+                    result = result.Add(reference, value);
+                }
+
+                return result;
+            }
+
+            // since we're enumerating a sorted collection, once we receive
+            // a mismatch it means we'll never again receive a match
+            var breakAfterFailure =
+                comparison == BinaryComparison.LessThanOrEqual ||
+                comparison == BinaryComparison.LessThanOrEqual;
+
+            var evaluator = comparison.GetEvaluator<TKey>();
+            foreach (var kvp in values)
+            {
+                if (evaluator(kvp.Key, reference))
+                {
+                    result = result.Add(kvp.Key, kvp.Value);
```

Comment:
these are immutable dictionaries. they maintain an AVL tree, so each operation is O(log(n)) if I recall correctly. 

That being said, during our chat I highlighted how I think we can definitely write some benchmarks around `OptionPositionCollection` and the functions he relies on, such as this guy. It will likely turn out that regular dictionaries might be better, or even a custom immutable type. The reason I liked the immutable types is for the reuse. One issue is since we're _very_ recursive, we need to make sure to pass new instances to each frame of the recursion. Consider drilling down a couple legs only to find out at the last leg that you can't find a match. If not using immutable collections, would need to keep track of all the removals and add them back in. With immutable collections we don't need to do anything since each match calls gets its own collection.